### PR TITLE
Add unit test for `updateAuthState:` in GIDGoogleUserTest

### PR DESCRIPTION
**Pull request by Alex-4-Git**
Our plan is to merge GIDAuthentication into GIDGoogleUser. This PR is to add more coverage to exiting GIDGoogleUser API and prepare for more changes in the coming PRs.

### CODE REVIEW

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 1]
```diff
@@ -91,4 +110,64 @@ - (void)testLegacyCoding {
 }
 #endif // TARGET_OS_IOS || TARGET_OS_MACCATALYST
 
+- (void) testUpdateAuthState {
+  GIDGoogleUser *user = [self googleUser];
+  XCTAssertEqualObjects(user.accessToken.tokenString, kAccessToken);
+  [self assertDate:user.accessToken.expirationDate equalTime:_accessTokenExpireTime];
```
Comment (mdmathias) [inline GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:116]: I think you're trying to make your tests DRY here, but sometimes that can add some mental overhead to reading and debugging failing tests. Consider removing this and [making your tests DAMP](https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html).

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 2]
```diff
@@ -91,4 +110,64 @@ - (void)testLegacyCoding {
...
+      [OIDTokenResponse testInstanceWithIDToken:idToken
+                                    accessToken:kAccessToken
+                                      expiresIn:accessTokenExpiresIn
+                                   tokenRequest:tokenRequest];
+  return [[GIDGoogleUser alloc]
+          initWithAuthState:[OIDAuthState testInstanceWithTokenResponse:tokenResponse]
+                profileData:nil];
+}
+
+- (NSString *)idToken {
```
Comment (mdmathias) [inline GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:147]: This and the getter below basically do the same thing with a different expiration time, right? Do you need these methods? If you're using these methods as a kind of named storage, consider using a property (or ivar) for each to make things clearer.

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 3]
```diff
@@ -91,4 +105,40 @@ - (void)testLegacyCoding {
...
+  XCTAssertEqualObjects(user.idToken.tokenString, idToken);
+  XCTAssertEqualWithAccuracy([user.idToken.expirationDate timeIntervalSinceReferenceDate],
+                             kIDTokenExpireTime, kTimeAccuracy);
+  
+  NSString *idTokenNew = [self idTokenWithExpireTime:kNewIDTokenExpireTime];
+  OIDAuthState *newAuthState = [OIDAuthState testInstanceWithIDToken:idTokenNew
+                                                         accessToken:kNewAccessToken
+                                               accessTokenExpireTime:kNewAccessTokenExpireTime];
+  
+  [user updateAuthState:newAuthState profileData:nil];
```
Comment (petea) [inline GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:128]: How about testing the case where `profileData` is provided?

Comment (Alex-4-Git) [reply GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m]: Done.

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 4]
```diff
@@ -34,10 +35,23 @@
 #import <AppAuth/OIDAuthState.h>
 #endif
 
+static NSString *const kNewAccessToken = @"new_access_token";
+static NSTimeInterval const kAccessTokenExpireTime = 442886100;
+static NSTimeInterval const kIDTokenExpireTime = 442886118;
+static NSTimeInterval const kNewAccessTokenExpireTime = 442886200;
```
Comment (mdmathias) [inline GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:41]: I don't think these constants are used outside of `testUpdateAuthState` below. Do you have plans to use them in future tests? If not, then maybe they don't need to be constants right now? 

This comment is driven by my concern that the constants don't really have meaningful names (or meaning in general) outside of `testUpdateAuthState`. 

You could create `NSTimeInterval`s within the method and give them good names in the new test method. Then, you could create new intervals with better names than "new" (e.g., something like `updatedAccessTokenExpireTime`) in the test method. 

[NSDate has a bunch of methods](https://developer.apple.com/documentation/foundation/date/1779636-timeintervalsince) that you can use to create time intervals.

PS I get the impression that you're following a pre-existing pattern in [GIDAuthenticationTest](https://github.com/google/GoogleSignIn-iOS/blob/main/GoogleSignIn/Tests/Unit/GIDAuthenticationTest.m), but IMHO it's not very readable there either.

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 5]
```diff
@@ -34,10 +35,23 @@
...
+static NSTimeInterval const kNewIDTokenExpireTime = 442886224;
+static NSTimeInterval const kTimeAccuracy = 10;
+
 @interface GIDGoogleUserTest : XCTestCase
 @end
 
-@implementation GIDGoogleUserTest
+@implementation GIDGoogleUserTest {
+  // Fake data used to generate the expiration date of the access token.
+  NSTimeInterval _accessTokenExpireTime;
```
[the remark this answers is not in the document]
Comment (Alex-4-Git) [reply GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m]: Removed.

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m (hunk 4 above)
Comment (Alex-4-Git) [reply GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m]: These value will be used in other test cases. But I think creating them inside the tests make the code easier to understand. Thanks for pointing it out.

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 6]
```diff
@@ -91,4 +110,64 @@ - (void)testLegacyCoding {
 }
 #endif // TARGET_OS_IOS || TARGET_OS_MACCATALYST
 
+- (void) testUpdateAuthState {
+  GIDGoogleUser *user = [self googleUser];
```
[the remark this answers is not in the document]
Comment (mdmathias) [reply GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m]: I see. Resolving.

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 7]
```diff
@@ -91,4 +95,48 @@ - (void)testLegacyCoding {
 }
 #endif // TARGET_OS_IOS || TARGET_OS_MACCATALYST
 
+- (void)testUpdateAuthState {
+  NSTimeInterval accessTokenExpireTime = [NSDate timeIntervalSinceReferenceDate];
+  NSTimeInterval idTokenExpireTime = [NSDate timeIntervalSinceReferenceDate] + 100;
```
Comment (mdmathias) [inline GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:100]: Do 100, 200, and 300 have special meaning, or are you just making sure that these times are different? I _think_ the time intervals should be different with each subsequent call to `[NSDate timeIntervalSinceReferenceDate]`. If so, maybe just drop the addition.

**Review by mdmathias — APPROVED** [review]

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m (hunk 7 above)
Comment (petea) [reply GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:100]: @Alex-4-Git could you respond to Matt's question here.

I'd also suggest marking comment threads as resolved when you have addressed them so as not to miss those that still need to be addressed.

Comment (Alex-4-Git) [reply GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m]: In the test we use a method `XCTAssertEqualWithAccuracy`. We want to make sure the difference between the expire times is larger than the accuracy so they would not be misused. 

I have used a constant kTimeIncrement to improve the readability.

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 8]
```diff
@@ -91,4 +98,40 @@ - (void)testLegacyCoding {
...
+  XCTAssertEqualObjects(user.idToken.tokenString, updatedIDToken);
+  XCTAssertEqualWithAccuracy([user.idToken.expirationDate timeIntervalSinceReferenceDate],
+                             updatedIDTokenExpireTime, kTimeAccuracy);
+  XCTAssertEqual(user.profile, updatedProfileData);
+}
+
+#pragma mark - Helpers
+
+- (NSString *)idTokenWithExpireTime:(NSTimeInterval)expireTime {
+  return [OIDTokenResponse idTokenWithSub:kUserID exp:@(expireTime + NSTimeIntervalSince1970)];
```
Comment (mdmathias) [inline GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:134]: Why does this method use `expireTime + NSTimeIntervalSince1970`? Removing `NSTimeIntervalSince1970` causes the tests to fail, which is confusing to me. 

Also, this method name suggests that it makes an ID token with the given `expireTime`, but what it does is create an ID token with the given expire time plus this extra increment. Maybe update the name and add a comment to clarify what it is doing and why.

Comment (Alex-4-Git) [reply GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:134]: This is the method we used to generate an id token in GIDAuthenticationTest. We can figure it out how it works offline and improve it in a separate PR.

Comment (petea) [reply GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:134]: > Why does this method use `expireTime + NSTimeIntervalSince1970`? Removing `NSTimeIntervalSince1970` causes the tests to fail, which is confusing to me.

The `NSTimeInterval` values generated by [`timeIntervalSinceReferenceDate`](https://developer.apple.com/documentation/foundation/nsdate/1408189-timeintervalsincereferencedate?language=objc) are relative to the reference date of Jan 1, 2001.  Here, [`NSTimeIntervalSince1970`](https://developer.apple.com/documentation/foundation/nstimeintervalsince1970?language=objc) is being used to convert it to a unix epoch value, which is the reference date used by the tokens.  We could probably simplify by using [`timeIntervalSince1970`](https://developer.apple.com/documentation/foundation/nsdate/1407504-timeintervalsince1970?language=objc) instead.

Comment (mdmathias) [reply GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:134]: Thanks for clarifying. I had a feeling that was the case. I agree that simplifying would be better

**Pull request merged**